### ISSUE TITLE
Insane amount of dns requests

### ISSUE THREAD

**Issue by CrazyWolf13** [louislam/uptime-kuma#4699]
### 📑 I have found these related issues/pull requests

None 

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/uptime-kuma/security/policy)

### Description

I'm getting a very very high amount of dns traffic because of uptime kuma, as it queries dns every 60 seconds for hundreds of domains, could you fix that by adding some sort of cache?

### 👟 Reproduction steps

Add a good amount of https domains, leave scrape time at default

### 👀 Expected behavior

Very little dns traffic, possibly once an hour, this would remove 60x dns traffix

### 😓 Actual Behavior

Every 60 seconds my dns server get's flooded with requests

### 🐻 Uptime-Kuma Version

1.23.11

### 💻 Operating System and Arch

I run Arch btw.

### 🌐 Browser

Google Chrome 124.0.6367.61

### 🖥️ Deployment Environment

tteck's helper scripts via LXC on proxmox: https://tteck.github.io/Proxmox/

**Comment by CommanderStorm** [comment]
You can activate [nscd (Name service caching daemon)](https://de.wikipedia.org/wiki/Name_service_caching_daemon) in the general settings.
By default, we respect the timeouts you configured in your DNS-Configuration:

![image](https://github.com/louislam/uptime-kuma/assets/26258709/a45f3176-dd10-4cd7-adf1-bcc4fa80d6a1)

**Comment by CrazyWolf13** [comment]
@CommanderStorm That's amazing, did not know something like that exists!

Update: I cannot view that im my settings !?

**Comment by CommanderStorm** [comment]
Overlooked that you have installed a script by a third party.
=> please consider either contacting them refering to this issue or use the docker container we provide at https://github.com/louislam/uptime-kuma/wiki/%F0%9F%94%A7-How-to-Install#-docker

The reason is that your installation option does
- not install and
- not configure nscd

We for now only enable this feature (and some others like "real browser") in containers to make the support effort somewhat manageable.

https://github.com/louislam/uptime-kuma/blob/29575343ca61461f0a07b334c688664d6e3ac47d/server/uptime-kuma-server.js#L450-L464

You can find the configuration file here:
https://github.com/louislam/uptime-kuma/blob/1.23.X/docker/etc/nscd.conf

**Comment by CrazyWolf13** [comment]
Okay I see, thanks for pointing that out, i'll let tteck know.

**Comment by CrazyWolf13** [comment]
> You can activate [nscd (Name service caching daemon)](https://de.wikipedia.org/wiki/Name_service_caching_daemon) in the general settings. By default, we respect the timeouts you configured in your DNS-Configuration:
> 
> ![image](https://private-user-images.githubusercontent.com/26258709/324260219-a45f3176-dd10-4cd7-adf1-bcc4fa80d6a1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTM3Mjc5NzQsIm5iZiI6MTcxMzcyNzY3NCwicGF0aCI6Ii8yNjI1ODcwOS8zMjQyNjAyMTktYTQ1ZjMxNzYtZGQxMC00Y2Q3LWFkZjEtYmNjNGZhODBkNmExLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNDIxVDE5Mjc1NFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExMGEwNzZhMmVhZjQ3ZTcwZWEwMWMzZGI5OTJhY2FmZDEzYzM1MTYxOTUwMmMwMzg4MGI5OTU3OTc0MGViMWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.HKjZ6pjyjzN--CqcTOM7_CMoU9-Dv_qYIX8cmOj-ym8)

I installed the service, injected your config and removed the part that checks for the variable in the shown code snippet, however at restart the js file was immediately rebuild, any hints on getting this to work?

------


As the owner is not really that happy minded to implement it, I tought I can use the dns cache for http(s) domains, however after activating still each and every request went to my pihole instance.

![image](https://github.com/louislam/uptime-kuma/assets/96661824/4b5b859a-6da3-4fa6-93bd-de1c1582def3)

**Comment by CommanderStorm** [comment]
> js file was immediately rebuild

It is actually just downloaded to prevent OOM errors. See [the third party script you use](https://raw.githubusercontent.com/tteck/Proxmox/main/ct/uptimekuma.sh): 

```
npm run download-dist &>/dev/null
```

Either you:
- pretend to be in docker via the `UPTIME_KUMA_IS_CONTAINER` environment variable and deal with missing dependencies
- use the installation opion (docker) from our wiki which includes all dependencys

**Comment by CrazyWolf13** [comment]
Is docker variable and installing the service, starting and adding your config did not solve, but I switched to docker, makes
things easier anyway.

**Comment by CrazyWolf13** [comment]
@CommanderStorm  Sorry to reopen this topic, but even after enabling the shown option oabove, and then after a while also enabling the other settings below the dns server setting also set to on i still get sooooo many dns requests, in 3 days I got like 75k just of uptime-kuma, is there any reason this cache is not working at all and the legacy cache also not?

**Comment by CommanderStorm** [comment]
25k/day for 100 monitors
=> 1 request/5m for 100 monitors
A TTL=5m is imo reasonable and no DNS resolver should not have a problem with this request load[^1].

Let me rephrase a previous comment (`By default, we respect the timeouts you configured in your DNS-Configuration`) via a quote from the nscd docs:

> for some name services (including specifically DNS) the TTL returned from the name service is used

=> Really difficult to gauge if this is excessive without additional data:

- What is the TTL of the DNS server resonses? (test just some, assume the rest are the same)
- How many monitors (total)?
- How many monitors with unique dns-hostnames (can be a guesstemate)?

For the legacy caching please see https://github.com/louislam/uptime-kuma/issues/3762
=> legacy caching is only active for some monitoring types and only for ipv4

[^1]: A resolver can handle hundreds to thousand concurrent connections and each resolution only takes few tenths of ms

**Comment by CrazyWolf13** [comment]
I have not specifically configured nscp, is that neccessary?

Okay yeah I also don't currently see it as a huge problem, it's just making reading dns logs and such things harded and it keeps my resolving quite high.

I have got 37 monitors where roundabout 30 are different domains other ip/ping/dns

TTL for my dns server is around 5-20ms depending on the domain.

**Comment by CommanderStorm** [comment]
> it's just making reading dns logs and such things harded

Use automated tooling such as anomaly detection or log based metrics via grafana loki and the likes.

> keeps my resolving quite high

Unclear what you mean

> TTL for my dns server is around 5-20ms

So basically TTL=0s => you have turned off DNS caching via this configuration
I did not know that DNS servers can be configured this low. Are you maybe confusing [RTT or latency with TTL](https://www.geeksforgeeks.org/difference-between-round-trip-time-rtt-and-time-to-live-ttl/)?

**Comment by CrazyWolf13** [comment]
> > it's just making reading dns logs and such things harded
> 
> Use automated tooling such as anomaly detection or log based metrics via grafana loki and the likes.
> 
> > keeps my resolving quite high
> 
> Unclear what you mean
> 
> > TTL for my dns server is around 5-20ms
> 
> So basically TTL=0s => you have turned off DNS caching via this configuration I did not know that DNS servers can be configured this low. Are you maybe confusing [RTT or latency with TTL](https://www.geeksforgeeks.org/difference-between-round-trip-time-rtt-and-time-to-live-ttl/)?

It's not the thing that there are too many dns requests, but the fact that there is still the same amout than before enabling the cache, has to has to be some sort of issue....


Yes you are right I confused it with time.

Not sure how I should be able to get the ttl. but I ran the following cmd found on the web from the uptime-kuma docker container which has my dns server as dns server.

```
user@-Tower:~$ dig +nocmd +noall +answer +ttlid A test.example.dev
test.example.dev.     0       IN      A       10.90.60.13

```

**Comment by CommanderStorm** [comment]
So your TTL is 0 => caching is disabled

**Comment by CrazyWolf13** [comment]
> So your TTL is 0 => caching is disabled

Well then how do I activate it?

The setting in the ui is turned on.

**Comment by CommanderStorm** [comment]
You can enable catching by allowing a higher TTL in your DNA settings.
